rfd/0095-okta-integration.md

```diff
@@ -0,0 +1,405 @@
+provider without access information, is expected to run every 10 minutes, as the list of
+applications and groups is not expected to change much. Additionally, the potential to have an
+adverse impact on the identity provider by overusing APIs is non-zero, so it's preferred to
+minimize the number of calls to identity provider APIs.
+
+```mermaid
+sequenceDiagram
+   participant IdPSvc as IdP Integration Service
+   participant IdP as Identity Provider
+   participant AppSvc as Application Access Service
+
+   loop Every 10 minutes
+      loop For each identity provider
+         IdPSvc->>IdP: Get list of applications
+         IdPSvc->>AppSvc: Synchronize applications in Teleport
+
+         IdPSvc->>IdP: Get list of groups
+
+         IdPSvc->>AppSvc: Synchronize groups in Teleport
+      end
+   end
+```
+
+Each provider will be iterated through and all applications will be retrieved and inserted into
+the Teleport database with a label matching the name of the provider. A new "external groups"
+object will be created to house external groups, and users will be able to belong to these
+groups.
+
+### User specific provider applications and groups
+
+Another Part of the identity provider integration service will be a user level synchronization
+service that runs every 2 minutes.  When a user logs in, if Teleport has been configured for
+external provider access, users will be this service, which will determine provider specific 
+applications and groups the user has access to. This information will be stored in our database
+and synchronized every 2 minutes. All logged in users will have this synchronization process
+run while they are logged in, but not while they are logged out.
+
+```mermaid
+sequenceDiagram
+   participant AuthSvc as Authentication Service
+   participant UserSync as User Synchronization Service
+   participant IdP as Identity Provider
+   participant DB as Database Access Service
+
+   AuthSvc->>+UserSync: Login registers user
+   loop Every 2 minutes
+      loop For each user
+         loop For each identity provider
+            UserSync->>IdP: Query list of applications for user
+            UserSync->>DB: Store results in database
+
+            UserSync->>IdP: Query list of groups for user
+            UserSync->>DB: Store results in database
+         end
+      end
+   end
+   UserSync->>-AuthSvc: Logout unregisters user
+```
+
+### Calculating user access to applications and groups
+
+The applications synchronized from external identity providers will be subject to regular
+Teleport RBAC rules, but further access is dependent on the external provider. The data stored
+from the user synchronization process will be joined to externally synchronized applications
+to determine whether a user has access to a particular application. The subsequent list of
+applications and joined user specific access data will be subject to the following logic to
+determine user access:
```

Comment:
Related to my RBAC question/suggestion above, I don't think we need to synchronize anything besides apps/groups themselves.